Mathlib/CategoryTheory/FiberedCategory/BasedCategory.lean

```diff
@@ -0,0 +1,274 @@
+
+/-- The composition of two based functors. -/
+@[simps!]
+def comp {𝒵 : BasedCategory.{v₄, u₄} 𝒮} (F : 𝒳 ⥤ᵇ 𝒴) (G : 𝒴 ⥤ᵇ 𝒵) : 𝒳 ⥤ᵇ 𝒵 where
+  toFunctor := F.toFunctor ⋙ G.toFunctor
+  w := by rw [Functor.assoc, G.w, F.w]
+
+@[simp]
+lemma comp_id (F : 𝒳 ⥤ᵇ 𝒴) : comp F (id 𝒴) = F :=
+  rfl
+
+@[simp]
+lemma id_comp (F : 𝒳 ⥤ᵇ 𝒴) : comp (id 𝒳) F = F :=
+  rfl
+
+@[simp]
+lemma comp_assoc {𝒵 : BasedCategory.{v₄, u₄} 𝒮} {𝒜 : BasedCategory.{v₅, u₅} 𝒮} (F : 𝒳 ⥤ᵇ 𝒴)
+    (G : 𝒴 ⥤ᵇ 𝒵) (H : 𝒵 ⥤ᵇ 𝒜) : comp (comp F G) H = comp F (comp G H) :=
+  rfl
+
+@[simp]
+lemma w_obj (F : 𝒳 ⥤ᵇ 𝒴) (a : 𝒳.obj) : 𝒴.p.obj (F.obj a) = 𝒳.p.obj a := by
+  rw [← Functor.comp_obj, F.w]
+
+instance (F : 𝒳 ⥤ᵇ 𝒴) (a : 𝒳.obj) : IsHomLift 𝒴.p (𝟙 (𝒳.p.obj a)) (𝟙 (F.obj a)) :=
+  IsHomLift.id (w_obj F a)
+
+section
+
+variable (F : 𝒳 ⥤ᵇ 𝒴) {R S : 𝒮} {a b : 𝒳.obj} (f : R ⟶ S) (φ : a ⟶ b)
+
+/-- For a based functor `F : 𝒳 ⟶ 𝒴`, then whenever an arrow `φ` in `𝒳` lifts some `f` in `𝒮`,
+then `F(φ)` also lifts `f`. -/
+instance preserves_isHomLift [IsHomLift 𝒳.p f φ] : IsHomLift 𝒴.p f (F.map φ) := by
+  apply of_fac 𝒴.p f (F.map φ) (Eq.trans (F.w_obj a) (domain_eq 𝒳.p f φ))
+    (Eq.trans (F.w_obj b) (codomain_eq 𝒳.p f φ))
+  rw [← Functor.comp_map, congr_hom F.w]
+  simpa using (fac 𝒳.p f φ)
+
+/-- For a based functor `F : 𝒳 ⟶ 𝒴`, and an arrow `φ` in `𝒳`, then `φ` lifts an arrow `f` in `𝒮`
+if `F(φ)` does. -/
+lemma isHomLift_map [IsHomLift 𝒴.p f (F.map φ)] : IsHomLift 𝒳.p f φ := by
+  apply of_fac 𝒳.p f φ  (F.w_obj a ▸ domain_eq 𝒴.p f (F.map φ))
+    (F.w_obj b ▸ codomain_eq 𝒴.p f (F.map φ))
+  simp [congr_hom F.w.symm, fac 𝒴.p f (F.map φ)]
+
+lemma isHomLift_iff : IsHomLift 𝒴.p f (F.map φ) ↔ IsHomLift 𝒳.p f φ :=
+  ⟨fun _ => isHomLift_map F f φ, fun _ => preserves_isHomLift F f φ⟩
+
+end
+
+end BasedFunctor
+
+
+/-- A `BasedNatTrans` between two `BasedFunctor`s is a natural transformation `α` between the
+underlying functors, such that for all `a : 𝒳`, `α.app a` lifts `𝟙 S` whenever `𝒳.p.obj a = S`. -/
+structure BasedNatTrans {𝒳 : BasedCategory.{v₂, u₂} 𝒮} {𝒴 : BasedCategory.{v₃, u₃} 𝒮}
+    (F G : 𝒳 ⥤ᵇ 𝒴) extends CategoryTheory.NatTrans F.toFunctor G.toFunctor where
+  isHomLift' : ∀ (a : 𝒳.obj), IsHomLift 𝒴.p (𝟙 (𝒳.p.obj a)) (toNatTrans.app a) := by aesop_cat
+
+namespace BasedNatTrans
+
+open BasedFunctor
+
+variable {𝒳 : BasedCategory.{v₂, u₂} 𝒮} {𝒴 : BasedCategory.{v₃, u₃} 𝒮}
+
+section
+
+variable {F G : 𝒳 ⥤ᵇ 𝒴} (α : BasedNatTrans F G)
+
+@[ext]
+lemma ext (β : BasedNatTrans F G) (h : α.toNatTrans = β.toNatTrans) : α = β := by
+  cases α; subst h; rfl
+
+instance app_isHomLift (a : 𝒳.obj) : IsHomLift 𝒴.p (𝟙 (𝒳.p.obj a)) (α.toNatTrans.app a) :=
+  α.isHomLift' a
+
+lemma isHomLift {a : 𝒳.obj} {S : 𝒮} (ha : 𝒳.p.obj a = S) :
+    IsHomLift 𝒴.p (𝟙 S) (α.toNatTrans.app a) := by
+  subst ha; infer_instance
+
+end
+
+/-- The identity natural transformation is a `BasedNatTrans`. -/
+@[simps!]
+def id (F : 𝒳 ⥤ᵇ 𝒴) : BasedNatTrans F F where
```

Comment:
```suggestion

initialize_simps_projections BasedNatTrans (+toNatTrans, -app)

/-- The identity natural transformation is a `BasedNatTrans`. -/
@[simps]
def id (F : 𝒳 ⥤ᵇ 𝒴) : BasedNatTrans F F where
```
And similarly, `simps!` has to be replaced by `simps` for `comp`. I am not a `simps` tactic guru, but what my suggestion does is that the `simps` attribute on definitions of `BasedNatTrans` will generate the `_toNatTrans` lemmas instead of the `_app` (which may be obtained using `simps!` if this were what we actually wanted).

Then, if `homCategory` is also tagged with `simps` instead of `simps!`, the two lemmas `id/comp_toNatTrans` that you added manually become unnecessary.

(You may try to make the same strategy work for `BasedFunctor`.)

Reply:
You missed the `initialize_simps_projections` part.

Reply:
Ah yes I did! So what is actually the point of this? I see that it allows me to write `@[simps]` on `id` instead of `@[simps toNatTrans]`, is that the main benefit?

Reply:
Yes. For all future declarations of based functors/nat. trans, `@[simps]` will work out of the box.